### PR TITLE
Add instrumentation for Armeria (server)

### PR DESCRIPTION
**Pull request by anuraaga**
Why?

- Armeria is my go-to framework so I want it instrumented!
- I wanted an example of a instrumented a framework that provides detailed logging, including timings, in the framework itself (so use its timestamps). Realized I forgot to add to end, will update it
- I hadn't written an instrumentation from scratch. It's a good experience :)

### CODE REVIEW

File: instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/OpenTelemetryService.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+  private OpenTelemetryService(HttpService delegate, ArmeriaServerTracer serverTracer) {
+    super(delegate);
+    this.serverTracer = serverTracer;
+  }
+
+  @Override
+  public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+    String route = route(ctx);
+    String spanName = route != null ? route : "HTTP " + req.method().name();
```
Comment (trustin) [inline instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/OpenTelemetryService.java:57]: Curious if we can utilize `ctx.log().serviceName()`

Comment (anuraaga) [reply instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/OpenTelemetryService.java:57]: Yeah let me check it in a followup PR where I add tests for armeria-grpc with it, for now I want to focus on the minimum HTTP behavior. Especially with auto instrumentation we might have some issue due to applying both instrumentations.

In the meantime maybe you can help with this one :)
 https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/756/files#diff-a63d5fb32f3eb41b3cde769b7807b2e6R87

File: instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          }
+        }
+      }
+    }
+
+    where:
+    path          | spanName     | method          | code
+    "/exact"      | "/exact"     | HttpMethod.GET  | 200
+    // TODO(anuraaga): Seems to be an Armeria bug not to have :objectId here
+    "/items/1234" | "/items/:"   | HttpMethod.POST | 200
```
Comment (minwoox) [inline instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy:88]: So you mean the `spanName` should be `/items/:itemsId`?

Comment (trustin) [inline instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy:88]: IIRC it's working as expected, although not very nice: https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/ParameterizedPathMapping.java#L64-L70

Armeria's `PathMapping` probably needs to provide a dedicated getter method for distributed tracing spans.

Comment (anuraaga) [reply instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy:88]: Yeah

Comment (anuraaga) [reply instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy:88]: Having a method like that to give a human readable representation of the route would be nice.

Comment (trustin) [reply instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy:88]: Created an issue for solving this better: https://github.com/line/armeria/issues/2932

File: instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import io.opentelemetry.auto.bootstrap.instrumentation.decorator.HttpServerTracer;
+import io.opentelemetry.context.propagation.HttpTextFormat.Getter;
+import io.opentelemetry.instrumentation.armeria.v0_99.internal.ContextUtil;
+import io.opentelemetry.trace.Tracer;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.net.URI;
+
+final class ArmeriaServerTracer
+    extends HttpServerTracer<HttpRequest, ServiceRequestContext, RequestContext> {
```
Comment (trask) [inline instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java:33]: it looks like we aren't using the `STORAGE` parameter, e.g. no calls to `getServerContext`

if that's the case, maybe good to be explicit about this (similar to `AkkaHttpServerTracer`)

```suggestion
    extends HttpServerTracer<HttpRequest, ServiceRequestContext, Void> {
```

File: instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/OpenTelemetryService.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    Span span = serverTracer.startSpan(req, ctx, spanName, requestStartTimeNanos);
+
+    // For non-recording spans, nothing special to do.
+    if (!span.isRecording()) {
+      try (Scope ignored = serverTracer.startScope(span, ctx)) {
+        return unwrap().serve(ctx, req);
+      }
+    }
+
+    ctx.log()
```
Comment (trask) [inline instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/OpenTelemetryService.java:72]: maybe wrap this in `span.isRecording()` and then don't need above condition/short-circuit?

File: instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+import com.linecorp.armeria.testing.junit4.server.ServerRule
+import io.opentelemetry.auto.test.InstrumentationSpecification
+import io.opentelemetry.auto.test.utils.TraceUtils
+import io.opentelemetry.trace.attributes.SemanticAttributes
+import spock.lang.Ignore
+import spock.lang.Shared
+
+import static io.opentelemetry.trace.Span.Kind.SERVER
+
+abstract class AbstractArmeriaServerTest extends InstrumentationSpecification {
```
Comment (trask) [inline instrumentation/armeria-0.99/testing/src/main/groovy/io/opentelemetry/instrumentation/armeria/v_0_99/AbstractArmeriaServerTest.groovy:34]: did you try extending `HttpServerTest`?

**Comment by trask** [comment]
Oh yes, and add it to the README!

File: instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java (hunk 3 above)
Comment (trask) [reply instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java:32]: ya, it's attached in `startScope`, but then never read after that

File: instrumentation/armeria-0.99/auto/armeria-0.99-auto.gradle [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+apply from: "$rootDir/gradle/instrumentation.gradle"
+
+// TODO(anuraaga): Move into instrumentation.gradle
+archivesBaseName = projectDir.parentFile.name
+
+muzzle {
+  pass {
+    group = "com.linecorp.armeria"
+    module = "armeria"
+    versions = "[0.99.8,)"
```
[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/armeria-0.99/auto/armeria-0.99-auto.gradle:14]: would it be weird to rename the module to armeria-1.0 now, just to avoid unnecessary maven artifact naming churn later? we can still doc the "correct" as of now version in README

File: README.md [hunk 7]
```diff
@@ -143,6 +143,7 @@ provide the path to a JAR file including an SPI implementation using the system
 | [Akka HTTP](https://doc.akka.io/docs/akka-http/current/index.html)                                                                    | 10.0+                          |
 | [Apache HttpAsyncClient](https://hc.apache.org/index.html)                                                                            | 4.0+                           |
 | [Apache HttpClient](https://hc.apache.org/index.html)                                                                                 | 2.0+                           |
+| [Armeria](https://armeria.dev)                                                                                                        | 0.98+                          |
```
Comment (trask) [inline README.md:146]: is this a typo? `0.99.8` was min version in the muzzle range

File: instrumentation/armeria-0.99/auto/armeria-0.99-auto.gradle (hunk 6 above)
[the remark this answers is not in the document]
Comment (anuraaga) [reply instrumentation/armeria-0.99/auto/armeria-0.99-auto.gradle]: Yeah think that makes sense

File: instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java (hunk 3 above)
Comment (anuraaga) [reply instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java]: Hmm I think I want to keep it like this, Armeria does provide a storage implementation, whether that's actually used or not is more a decision of `HttpServerTracer` than the armeria instrumentation from what I can tell.

Comment (trask) [reply instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java]: > whether that's actually used or not is more a decision of `HttpServerTracer` than the armeria instrumentation from what I can tell

so, confusingly, this is not the case. it's up to the individual instrumentation to use what's in `STORAGE` if they need it for propagation

it seems likely we will need it in the armeria instrumentation in the future, e.g. for propagating context to some async handlers, so we can keep it, though maybe a comment about it not being used currently?

@iNikem please correct me if I'm missing something about the `STORAGE` param (either current or future plans).

Comment (trask) [reply instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java]: oh yes, and i opened #791 to document the `STORAGE` param, it is confusing when/if to use it

Comment (anuraaga) [reply instrumentation/armeria-0.99/library/src/main/java/io/opentelemetry/instrumentation/armeria/v0_99/server/ArmeriaServerTracer.java]: Ok less complexity makes sense. I'll revisit after instrumenting client and introducing some more complex async scenarios.
